src/rt/rwlock.rs

```diff
@@ -0,0 +1,269 @@
+    seq_cst: bool,
+
+    /// `Some` when the rwlock is in the locked state.
+    /// The `thread::Id` references the thread that currently holds the rwlock.
+    lock: Option<RwLockType>,
+
+    /// Tracks write access to the rwlock.
+    last_access: Option<Access>,
+
+    /// Causality transfers between threads
+    synchronize: Synchronize,
+}
+
+impl RwLock {
+    /// Common RwLock function
+    pub(crate) fn new(seq_cst: bool) -> RwLock {
+        super::execution(|execution| {
+            let obj = execution.objects.insert_rwlock(State {
+                seq_cst,
+                lock: None,
+                last_access: None,
+                synchronize: Synchronize::new(execution.max_threads),
+            });
+
+            RwLock { obj }
+        })
+    }
+
+    fn get_state<'a>(&self, objects: &'a mut object::Store) -> &'a mut State {
+        self.obj.rwlock_mut(objects).unwrap()
+    }
+}
+
+impl RwLock {
+    /// Read RwLock functions
+    pub(crate) fn acquire_read_lock(&self) {
+        self.obj.branch_acquire(self.is_write_locked());
+        assert!(
+            self.post_acquire_read_lock(),
+            "expected to be able to acquire read lock"
+        );
+    }
+
+    pub(crate) fn try_acquire_read_lock(&self) -> bool {
+        self.obj.branch_opaque();
+        self.post_acquire_read_lock()
+    }
+
+    pub(crate) fn release_read_lock(&self) {
+        super::execution(|execution| {
+            let state = self.get_state(&mut execution.objects);
+
+            state.lock = None;
+
+            state
+                .synchronize
+                .sync_store(&mut execution.threads, Release);
+
+            if state.seq_cst {
+                execution.threads.seq_cst();
+            }
+
+            let thread_id = execution.threads.active_id();
+
+            for (id, thread) in execution.threads.iter_mut() {
+                if id == thread_id {
+                    continue;
+                }
+
+                let obj = thread
+                    .operation
+                    .as_ref()
+                    .map(|operation| operation.object());
+
+                if obj == Some(self.obj) {
+                    thread.set_runnable();
+                }
+            }
+        });
+    }
+
+    fn post_acquire_read_lock(&self) -> bool {
+        super::execution(|execution| {
+            let state = self.get_state(&mut execution.objects);
+            let thread_id = execution.threads.active_id();
+
+            let thread_set = match &state.lock {
+                None => {
+                    let mut threads: HashSet<thread::Id> = HashSet::new();
+                    threads.insert(thread_id);
+                    threads
+                }
+                Some(RwLockType::Read(current)) => {
+                    let mut threads: HashSet<thread::Id> = HashSet::new();
+                    threads.extend(current);
+                    threads.insert(thread_id);
+                    threads
```

Comment:
Is there a reason we create a new `HashSet` here, rather than something like this:
```rust
let thread_set = match state.lock {
    // ...
    Some(RwLockType::Read(ref mut current)) => {
        current.insert(thread_id);
        current
    },
    // ...
}

Reply:
I didn't go down that route because it quickly unraveled into needing to specify lifetime parameters everywhere and I am not adept enough at Rust to do that... with grace. I'll dig into it though and see if I can make it work.

Reply:
Hmm, if it isn't easily changed, I don't think this is a blocker, it just seems kind of inefficient...